src/uct/base/uct_iface.c

```diff
@@ -548,3 +549,44 @@ ucs_config_field_t uct_iface_config_table[] = {
 
   {NULL}
 };
+
+ucs_status_t uct_base_ep_am_short_iov(uct_ep_h ep, uint8_t id, const uct_iov_t *iov,
+                                      size_t iovcnt)
+{
+    uint64_t header = 0;
+    size_t offset   = 0;
+    unsigned length;
+    void *buffer;
+    size_t iov_it;
+    void *payload;
+    unsigned payload_length;
+    ucs_status_t status;
+
+    length = uct_iov_total_length(iov, iovcnt);
+    if (length > UCS_ALLOCA_MAX_SIZE) {
+        buffer = malloc(length);
+    } else {
+        buffer = ucs_alloca(length);
+    }
+
+    for (iov_it = 0; iov_it < iovcnt; ++iov_it) {
+        memcpy(UCS_PTR_BYTE_OFFSET(buffer, offset), iov[iov_it].buffer,
+               iov[iov_it].length);
+        offset += iov[iov_it].length;
+    }
+
+    memcpy(&header, buffer, ucs_min(length, sizeof(header)));
+    if (length > sizeof(header)) {
+        payload        = UCS_PTR_TYPE_OFFSET(buffer, header);
+        payload_length = length - sizeof(header);
+    } else {
+        payload        = NULL;
+        payload_length = 0;
+    }
```

Comment:
seems header is not needed (and no need to copy to it). Can allocate `length + sizeof(header)` and then call am short like below (short pseudo code)
```
uct_ep_am_short(ep, id, buffer, buffer + sizeof(hdr), payload_length - sizeof(hdr));
```

Reply:
agree